src/lib/jobrunner.rs

```diff
@@ -0,0 +1,409 @@
+use whoami::{hostname, username};
+
+use crate::{queue::QueueJob, Snare};
+
+const READBUF: usize = 8 * 1024; // Bytes. Should be >= PIPE_BUF for performance reasons.
+const WAIT_TIMEOUT: i32 = 1; // Seconds.
+const RECHECK_TIMEOUT: i32 = 5; // Seconds.
+const DEFAULT_TIMEOUT: i32 = 60; // Seconds. How often to check the queue "just in case?
+
+struct JobRunner {
+    snare: Arc<Snare>,
+    /// The running jobs, with `0..num_running` entries.
+    running: Vec<Option<Job>>,
+    /// How many `Some` entries are there in `self.running`?
+    num_running: usize,
+    /// The running jobs, with `0..2 *num_running + 1` entries. Each pair of entries are (stderr,
+    /// stdout) for the corresponding `running` `Job` (i.e. `running[0]` has its stderr entry at
+    /// `pollfds[0]` and stdout entry at `pollfds[1]`). The `+ 1` entry is the event file
+    /// descriptor that allows the HTTP server thread to wake up the JobRunner thread.
+    pollfds: Vec<PollFd>,
+}
+
+impl JobRunner {
+    fn new(snare: Arc<Snare>) -> Result<Self, Box<dyn Error>> {
+        if snare.config.maxjobs == std::usize::MAX {
+            eprintln!(
+                "The maximum number of possible jobs is {}",
+                std::usize::MAX - 1
+            );
+            process::exit(1);
+        }
+        let mut running = Vec::with_capacity(snare.config.maxjobs);
+        running.resize_with(snare.config.maxjobs, || None);
+        let mut pollfds = Vec::with_capacity(snare.config.maxjobs * 2 + 1);
+        pollfds.resize_with(snare.config.maxjobs * 2 + 1, || {
+            PollFd::new(-1, PollFlags::empty())
+        });
+        Ok(JobRunner {
+            snare,
+            running,
+            num_running: 0,
+            pollfds,
+        })
+    }
+
+    fn attend(&mut self) {
+        self.update_pollfds();
+        // `check_queue` serves two subtly different purposes:
+        //   * Has the event pipe told us there are new jobs in the queue?
+        //   * Are there jobs in the queue from a previous round that we couldn't run yet?
+        let mut check_queue = false;
+        // How many jobs have closed stderr/stdout and which we're just waiting to terminate?
+        let mut num_waiting = 0;
+        // A scratch buffer used to read from files.
+        let mut buf = Box::new([0; READBUF]);
+        loop {
+            let timeout = if num_waiting > 0 {
+                WAIT_TIMEOUT
+            } else if check_queue {
+                RECHECK_TIMEOUT
+            } else {
+                DEFAULT_TIMEOUT
+            };
+            poll(&mut self.pollfds, timeout * 1000).ok();
+
+            // See if any of our active jobs have events. Knowing when a pipe is actually closed is
+            // surprisingly hard. https://www.greenend.org.uk/rjk/tech/poll.html has an interesting
+            // suggestion which we adapt slightly here.
+            for i in 0..self.snare.config.maxjobs {
+                // stderr
```

Comment:
partial comment left in?

Reply:
Maybe not though, because there's another similar one below for stdout.

So what about stderr?

Nit: other comments in the file have a full stop.

Reply:
There are two parts to that `for` loop: the bit that handles stderr and the bit that handles stdout. They're easy to confuse, hence the minor comment.